core/src/main/scala/org/apache/spark/rpc/ActionScheduler.scala

```diff
@@ -0,0 +1,207 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rpc
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.{SynchronousQueue, TimeUnit, ThreadPoolExecutor}
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.ExecutionContext
+import scala.util.control.NonFatal
+
+import io.netty.util.{Timeout, TimerTask, HashedWheelTimer}
+
+import org.apache.spark.util.Utils
+import org.apache.spark.{Logging, SparkConf}
+
+/**
+ * It's very common that executing some actions in other threads to avoid blocking the event loop
+ * in a RpcEndpoint. [[ActionScheduler]] is designed for such use cases.
+ */
+private[spark] trait ActionScheduler {
```

Comment:
`executeIOAction` is used to run some blocking but non-CPU actions, such as reading files or sending/receiving via the network. `executeCPUAction` is used to run some CPU actions, such as compress/decompress, or serialize/deserialize.